server/lease/lessor.go

```diff
@@ -326,6 +326,12 @@ func (le *lessor) Revoke(id LeaseID) error {
 		le.mu.Unlock()
 		return ErrLeaseNotFound
 	}
+
+	// We shouldn't delete the lease inside the transaction lock, otherwise
+	// it may lead to deadlock with Grant or Checkpoint operations, which
+	// acquire the le.mu firstly and then the batchTx lock.
```

Comment:
I agree that there is a chance of deadlock here, however I'm worried about removing lease from leaseMap and releasing lock before db operation is executed. This might be totally unfounded worry as backend operation as flushed only once 5 seconds, however I need to think a little on it. Have you thought on it more?

Reply:
It should be OK. The` le.mu` is used to protect the data structures inside `lessor`, and the backend transaction is protected by separate transaction lock.  Just as I mentioned previously, there is no reason to hold the `le.mu` during the transaction. 

Please note that during `Grant` and `Checkpoint`, lessor may also persist the lease, but both of them will require the transaction lock before persisting the lease data.  So it's safe.

Reply:
For my understanding, 

before the change 

```
Grant is (called Process 1)
1. acquires lessor.mu R1 (called Resource 1)
2. acquires backend tx R2
3. releases backend tx R2
4. releases lessor.mu R1

Revoke P2
1. acquires lessor.mu R1
2. releases lessor.mu R1
3. acquires backend tx R2
4. acquires lessor.mu R1
5. releases backend tx R2
6. releases lessor.mu R1
```

There is a potential deadlock for sure
- P1 is holding R1 and wait to acquire R2
- P2 is holding R2 and wait to acquire R1

Why not we move P2 step 4 to happen after step 5? Namely `delete(le.leaseMap, id)` after `txn.End()` which releases the backend transaction.